vertical-pod-autoscaler/pkg/utils/vpa/capping.go

```diff
@@ -176,23 +176,43 @@ func applyVPAPolicy(recommendation apiv1.ResourceList, policy *vpa_types.Contain
 
 func applyVPAPolicyForContainer(containerName string,
 	containerRecommendation *vpa_types.RecommendedContainerResources,
-	policy *vpa_types.PodResourcePolicy) (*vpa_types.RecommendedContainerResources, error) {
+	policy *vpa_types.PodResourcePolicy,
+	globalMaxAllowed apiv1.ResourceList) (*vpa_types.RecommendedContainerResources, error) {
 	if containerRecommendation == nil {
 		return nil, fmt.Errorf("no recommendation available for container name %v", containerName)
 	}
 	cappedRecommendations := containerRecommendation.DeepCopy()
-	// containerPolicy can be nil (user does not have to configure it).
 	containerPolicy := GetContainerResourcePolicy(containerName, policy)
-	if containerPolicy == nil {
-		return cappedRecommendations, nil
-	}
 
 	process := func(recommendation apiv1.ResourceList) {
 		for resourceName, recommended := range recommendation {
-			cappedToMin, _ := maybeCapToPolicyMin(recommended, resourceName, containerPolicy)
-			recommendation[resourceName] = cappedToMin
-			cappedToMax, _ := maybeCapToPolicyMax(cappedToMin, resourceName, containerPolicy)
-			recommendation[resourceName] = cappedToMax
+			// containerPolicy can be nil (user does not have to configure it).
+			if containerPolicy != nil {
+				cappedToMin, _ := maybeCapToPolicyMin(recommended, resourceName, containerPolicy)
+				recommendation[resourceName] = cappedToMin
+
+				maxAllowed := containerPolicy.MaxAllowed
+				if globalMaxAllowed != nil {
+					if maxAllowed == nil {
+						maxAllowed = globalMaxAllowed
+					} else {
+						// Set resources from the global maxAllowed if the VPA maxAllowed is missing them.
+						for resourceName, quantity := range globalMaxAllowed {
+							if _, ok := maxAllowed[resourceName]; !ok {
+								maxAllowed[resourceName] = quantity
+							}
+						}
+					}
+				}
+
+				cappedToMax, _ := maybeCapToMax(cappedToMin, resourceName, maxAllowed)
+				recommendation[resourceName] = cappedToMax
+			} else {
+				if globalMaxAllowed != nil {
+					cappedToMax, _ := maybeCapToMax(recommended, resourceName, globalMaxAllowed)
+					recommendation[resourceName] = cappedToMax
+				}
+			}
 		}
 	}
```

Comment:
Would this be a little more simpler to read?
```
	process := func(recommendation apiv1.ResourceList) {
		for resourceName, recommended := range recommendation {
			var maxAllowed apiv1.ResourceList
                        // containerPolicy can be nil (user does not have to configure it).
			if containerPolicy != nil {
				cappedToMin, _ := maybeCapToPolicyMin(recommended, resourceName, containerPolicy)
				recommendation[resourceName] = cappedToMin
				maxAllowed = containerPolicy.MaxAllowed
			}

			if globalMaxAllowed != nil {
				if maxAllowed == nil {
					maxAllowed = globalMaxAllowed
				} else {
					for resourceName, quantity := range globalMaxAllowed {
						if _, ok := maxAllowed[resourceName]; !ok {
							maxAllowed[resourceName] = quantity
						}
					}
				}
			}

			cappedToMax, _ := maybeCapToMax(recommendation[resourceName], resourceName, maxAllowed)
			recommendation[resourceName] = cappedToMax
		}
	}
```

Reply:
Agree. I addressed the suggestion in https://github.com/kubernetes/autoscaler/pull/7560/commits/04e33401fe93d83320414e87f30570cdc24586de.